### ISSUE TITLE
UCI bug

### ISSUE THREAD

**Issue by tissatussa** [JoostHouben/Euwe-chess-engine#1]
I managed to compile (with clang++-17) your v1.0.1 source on Linux, but only after changing some code to suit typical c++23 constructs.
Then the engine seems to have trouble with the UCI protocol :

```
$ ./euwe-engine
uci
id name singular-early-exit
id author Joost Houben
option name Hash type spin default 0 min 0 max 1048576
option name move_overhead_ms type spin default 100 min 0 max 10000
uciok
ucinewgame
isready
readyok
position startpos
[DEBUG] Error: Unrecognized token 'startpos'. Expected 'moves'.
quit
```

Same when giving a FEN position to start a game :

```
$ ./euwe-engine
uci
id name singular-early-exit
id author Joost Houben
option name Hash type spin default 0 min 0 max 1048576
option name move_overhead_ms type spin default 100 min 0 max 10000
uciok
ucinewgame
isready
readyok
position fen rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq - 0 1
[DEBUG] Error: Unrecognized token '1'. Expected 'moves'.
quit
```

I guess this is a bug and easy to fix ?

**Comment by JoostHouben** [comment]
Thanks for trying to build the engine on Linux, I'm glad that worked without too much trouble!

The error message is admittedly a bit confusing, but this is not a bug; the UCI protocol requires that every 'position' command ends with the token 'moves':

> `position [fen <fenstring> | startpos ]  moves <move1> .... <movei>`

Note how in that specification the `moves` token is not marked as optional (it would be put in square brackets if it was).

If you include the `moves` token things work as expected:

```
uci
id name 7th-rank-extension
id author Joost Houben
option name Hash type spin default 16 min 0 max 1048576
option name move_overhead_ms type spin default 100 min 0 max 10000
uciok
ucinewgame
isready
readyok
debug on
info string Debug mode enabled: true
position startpos moves
info string FEN: rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1
[DEBUG] Position:
  .-------------------------------.
8 | r*| n | b | q | k | b | n | r*|
  |---+---+---+---+---+---+---+---|
7 | p | p | p | p | p | p | p | p |
  |---+---+---+---+---+---+---+---|
6 |   |   |   |   |   |   |   |   |
  |---+---+---+---+---+---+---+---|
5 |   |   |   |   |   |   |   |   |
  |---+---+---+---+---+---+---+---|
4 |   |   |   |   |   |   |   |   |
  |---+---+---+---+---+---+---+---|
3 |   |   |   |   |   |   |   |   |
  |---+---+---+---+---+---+---+---|
2 | P | P | P | P | P | P | P | P |
  |---+---+---+---+---+---+---+---|
1 | R*| N | B | Q | K | B | N | R*| w
  '-------------------------------'
    a   b   c   d   e   f   g   h
```

Nevertheless, I agree with you that it's a bit cumbersome to have to type 'moves' when there are no moves to parse if you're interacting with the CLI manually. To improve the ergonomics of the CLI I have patched the UCI parser to accept `position` commands without the `moves` token (if no moves are specified).

**Comment by tissatussa** [comment]
No, you're mistaken. In the CuteChess GUI Debug Pane you can see the UCI protocol acts like this (2 UCI engines), for startposition and FEN :

```
<Alexander v1.3 default MCTS(27): Alexander 1.3 by K. Kiniama, A. Manzo and Stockfish developers (see AUTHORS file)
<4KU v5.0(26): readyok
<Alexander v1.3 default MCTS(27): id name Alexander 1.3
<Alexander v1.3 default MCTS(27): id author K. Kiniama, A. Manzo and Stockfish developers (see AUTHORS file)
<Alexander v1.3 default MCTS(27): option name Debug Log File type string default 
<Alexander v1.3 default MCTS(27): option name Threads type spin default 1 min 1 max 1024
<Alexander v1.3 default MCTS(27): option name Hash type spin default 16 min 1 max 33554432
(--more-options--)
<Alexander v1.3 default MCTS(27): uciok
>Alexander v1.3 default MCTS(27): setoption name Debug Log File
>Alexander v1.3 default MCTS(27): setoption name Threads value 2
>Alexander v1.3 default MCTS(27): setoption name Hash value 128
(--more-setoptions--)
>Alexander v1.3 default MCTS(27): isready
<Alexander v1.3 default MCTS(27): readyok
>4KU v5.0(26): ucinewgame
>4KU v5.0(26): position startpos
>Alexander v1.3 default MCTS(27): ucinewgame
>Alexander v1.3 default MCTS(27): setoption name Ponder value false
>Alexander v1.3 default MCTS(27): position startpos
>4KU v5.0(26): isready
<4KU v5.0(26): readyok
>4KU v5.0(26): go wtime 303000 btime 303000 winc 3000 binc 3000
<4KU v5.0(26): info depth 1 score cp 32 lowerbound time 1 nodes 18 nps 18000 pv b1c3 b8c6
<4KU v5.0(26): info depth 1 score cp 29 time 1 nodes 38 nps 38000 pv b1c3
(--more-info-lines--)
<4KU v5.0(26): bestmove e2e4
>Alexander v1.3 default MCTS(27): position startpos moves e2e4
>Alexander v1.3 default MCTS(27): isready
<Alexander v1.3 default MCTS(27): readyok
>Alexander v1.3 default MCTS(27): go wtime 294532 btime 303000 winc 3000 binc 3000
<Alexander v1.3 default MCTS(27): info depth 1 seldepth 2 multipv 1 score cp 20 wdl 65 919 16 nodes 41 nps 41000 hashfull 0 tbhits 0 time 1 pv b8c6
<Alexander v1.3 default MCTS(27): info depth 2 seldepth 3 multipv 1 score cp 3 wdl 37 934 29 nodes 1782 nps 891000 hashfull 0 tbhits 0 time 2 pv g8f6 f2f3
(--more-info-lines--)
<Alexander v1.3 default MCTS(27): info depth 30 currmove f7f6 currmovenumber 20
<Alexander v1.3 default MCTS(27): info depth 30 seldepth 55 multipv 1 score cp -9 wdl 23 932 45 nodes 124449112 nps 2697791 hashfull 1000 tbhits 0 time 46130 pv e7e6 d2d4 d7d5 e4d5 e6d5 g1f3 g8f6 f1d3 f8d6 d1e2 d6e7 e1g1 e8g8 c1e3 f8e8 c2c4 b8c6 a2a3 d5c4 d3c4 c8g4 b1c3 c6d4 c4f7 g8f7 e2c4 g4e6 f3g5 f7f8 c4d4 e6b3 d4f4 f8g8 g5e4 e7d6 e4d6 d8d6 f4d6 c7d6 a1c1
<Alexander v1.3 default MCTS(27): bestmove e7e6 ponder d2d4
>4KU v5.0(26): position startpos moves e2e4 e7e6
>4KU v5.0(26): isready
<4KU v5.0(26): readyok
>4KU v5.0(26): go wtime 294532 btime 259869 winc 3000 binc 3000
<4KU v5.0(26): info depth 1 score cp 32 lowerbound time 1 nodes 1 nps 1000 pv g1f3
<4KU v5.0(26): info depth 1 score cp 33 time 1 nodes 31 nps 31000 pv b1c3
(--etc--)

--------------------------------------

<4KU v5.0(28): readyok
<Alexander v1.3 default MCTS(29): id name Alexander 1.3
<Alexander v1.3 default MCTS(29): id author K. Kiniama, A. Manzo and Stockfish developers (see AUTHORS file)
<Alexander v1.3 default MCTS(29): option name Debug Log File type string default 
<Alexander v1.3 default MCTS(29): option name Threads type spin default 1 min 1 max 1024
<Alexander v1.3 default MCTS(29): option name Hash type spin default 16 min 1 max 33554432
(--more-options--)
>Alexander v1.3 default MCTS(29): isready
<Alexander v1.3 default MCTS(29): readyok
>4KU v5.0(28): ucinewgame
>4KU v5.0(28): position fen rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq - 0 1
>Alexander v1.3 default MCTS(29): ucinewgame
>Alexander v1.3 default MCTS(29): setoption name Ponder value false
>Alexander v1.3 default MCTS(29): position fen rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq - 0 1
>Alexander v1.3 default MCTS(29): isready
<Alexander v1.3 default MCTS(29): readyok
>Alexander v1.3 default MCTS(29): go wtime 303000 btime 303000 winc 3000 binc 3000
<Alexander v1.3 default MCTS(29): info depth 1 seldepth 2 multipv 1 score cp 0 wdl 33 934 33 nodes 20 nps 10000 hashfull 0 tbhits 0 time 2 pv e7e6
<Alexander v1.3 default MCTS(29): info depth 2 seldepth 3 multipv 1 score cp 0 wdl 32 935 33 nodes 873 nps 436500 hashfull 0 tbhits 0 time 2 pv g8f6 b1c3
(--more-info-lines--)
<Alexander v1.3 default MCTS(29): info depth 30 currmove f7f5 currmovenumber 20
<Alexander v1.3 default MCTS(29): info depth 33 seldepth 56 multipv 1 score cp 0 wdl 33 934 33 nodes 124743163 nps 2564673 hashfull 1000 tbhits 0 time 48639 pv g8f6 b1c3 d7d5 e4d5 f6d5 f1c4 c7c6 g1f3 d5c3 b2c3 b8d7 e1g1 e7e6 g2g3 f8d6 d2d4 e8g8 d1d3 d7f6 c4b3 d8a5 c3c4 a5h5 f1e1 f8e8 e1e2 f6g4 c1b2 c8d7 a1e1 h7h6 e2d2 a8d8 h2h4 b7b6 a2a3
<Alexander v1.3 default MCTS(29): bestmove g8f6 ponder b1c3
>4KU v5.0(28): position fen rnbqkbnr/pppppppp/8/8/4P3/8/PPPP1PPP/RNBQKBNR b KQkq - 0 1 moves g8f6
>4KU v5.0(28): isready
<4KU v5.0(28): readyok
>4KU v5.0(28): go wtime 303000 btime 257360 winc 3000 binc 3000
<4KU v5.0(28): info depth 1 score cp 14 time 0 nodes 32 pv e4e5
<4KU v5.0(28): info depth 2 score cp 6 time 1 nodes 159 nps 159000 pv b1c3 b8c6
(--etc--)
```

> I'm glad that worked without too much trouble!

Well, i struggled a lot with it .. those typical C++23 constructs almost gave me a headache because on my Linux (Xubuntu 22.04) they seem only partially implemented .. eg. `std::ranges::views::join_with()` isn't known, and also `std::println()` .. i think the latter is a basic function, so it's weird .. i had to replace them by alternative constructs (with the help of ChatGPT) .. also i had to create a special Makefile because i don't use VS.

![linux-rules](https://github.com/user-attachments/assets/c1651fbd-cda5-41e2-829b-ee122e71968a)

@@ when will all you Window$ users discover Linux is a blessing @@

**Comment by AlexBrunetti** [comment]
Although the official UCI protocol documentation explicitly states that moves is not optional, common practice among chess engines and game interfaces shows that almost all of them accept the position command without the moves parameter. This behavior is influenced by the design and compatibility of the engines rather than a formal declaration in the UCI protocol syntax.

Alex

**Comment by tissatussa** [comment]
AFAIK no _official_ UCI protocol exists.

**Comment by JoostHouben** [comment]
Thanks for the additional information. I guess the engine still worked with UIs that omit the 'moves' token because it falls back to simply using the starting position if it can't parse the 'position' command.

As I said, the code has been patched to accept the variant without 'moves', so this should now be fixed. Thanks for reporting!

**Comment by AlexBrunetti** [comment]
> AFAIK no _official_ UCI protocol exists.

Well, I think it depends on how we define "official." For me, the original post from its inventor serves as the official reference.

Alex

**Comment by tissatussa** [comment]
@JoostHouben No, you didn't fix it .. CuteChess GUI gives an error :

```
<Euwe v1.0.1(34): id name 7th-rank-extension
<Euwe v1.0.1(34): id author Joost Houben
<Euwe v1.0.1(34): option name Hash type spin default 16 min 0 max 1048576
<Euwe v1.0.1(34): option name move_overhead_ms type spin default 100 min 0 max 10000
<Euwe v1.0.1(34): uciok
>Euwe v1.0.1(34): setoption name Hash value 128
>Euwe v1.0.1(34): setoption name move_overhead_ms value 100
>Euwe v1.0.1(34): isready
<4KU v5.0(35): readyok
<Euwe v1.0.1(34): [DEBUG] Option 'Hash' was set to '128'.
<Euwe v1.0.1(34): [DEBUG] Option 'move_overhead_ms' was set to '100'.
<Euwe v1.0.1(34): readyok
>Euwe v1.0.1(34): ucinewgame
>Euwe v1.0.1(34): position startpos
>4KU v5.0(35): ucinewgame
>4KU v5.0(35): position startpos
>Euwe v1.0.1(34): isready
<Euwe v1.0.1(34): [DEBUG] Warning: did not find token 'moves' in position command. Position commands without 'moves' are accepted as a non-standard extension when no move list is specified. If a move list was specified but the 'moves' token was omitted, the command may be parsed incorrectly.
<Euwe v1.0.1(34): readyok
>Euwe v1.0.1(34): go wtime 303000 btime 303000 winc 3000 binc 3000
<Euwe v1.0.1(34): [DEBUG] Time budget: 10400 ms
>4KU v5.0(35): isready
<4KU v5.0(35): readyok
```

Btw. i copied all changed code parts into my (Linux) version, using Meld.
Compilation went well, i think i didn't make a mistake here ..

You should test in CuteChess GUI to see if it works in practice.

And don't close an Issue like this too soon : you might think it's solved, but some related issues may not ..

**Comment by tissatussa** [comment]
@Alex you're right, we all use the April 2006 version, i think that's the one you're refering to.

**Comment by JoostHouben** [comment]
@tissatussa Anything labeled [DEBUG] should not be sent to the GUI. In the original source code anything labeled like that is sent to the stderr stream, which is not connected to the GUI (only stdout is). Perhaps you made a mistake in converting the `println` statements to work around the C++23 requirement and accidentally made the debug statements print to stdout.

**Comment by tissatussa** [comment]
> Anything labeled [DEBUG] should not be sent to the GUI (..) Perhaps you made a mistake in converting the println statements to work around the C++23 requirement and accidentally made the debug statements print to stdout.

yes, i might have, thanks, i will look into it.
nevertheless, it can't play a game in CuteChess (and this should be our reference).

**Comment by JoostHouben** [comment]
It successfully plays games in CuteChess (and other GUIs and interfaces) when built from the original source on Windows. So I don't believe there's an open compatibility issue with the original source.

Linux is currently not a supported platform, though I do plan to support it in the future. I certainly encourage you to make your own modifications to run it on Linux, but it's not a scenario I can presently debug or support.

**Comment by AlexBrunetti** [comment]
> @alex you're right, we all use the April 2006 version, i think that's the one you're refering to.

I'm referring the November 28, 2000 original post. But also in the April 15, 2004 extension by the author and the April 2006 update from Shredder, "moves" is showed as not optional.

Alex

**Comment by tissatussa** [comment]
@JoostHouben well, i must have made some fault in the code conversion .. this is terminal output :

```
$ ./euwe-engine-v1.0.1-compiled-HP 
uci
id name 7th-rank-extension
id author Joost Houben
option name Hash type spin default 16 min 0 max 1048576
option name move_overhead_ms type spin default 100 min 0 max 10000
uciok
ucinewgame
isready
readyok
position startpos
[DEBUG] Warning: did not find token 'moves' in position command. Position commands without 'moves' are accepted as a non-standard extension when no move list is specified. If a move list was specified but the 'moves' token was omitted, the command may be parsed incorrectly.
go wtime 70000 btime 70000 winc 10000 binc 10000    
[DEBUG] Time budget: 11400 ms
euwe-engine-v1.0.1-compiled-HP: chess-engine-lib/StackOfVectors.h:370: const T &StackVector<Move>::operator[](int) const [T = Move]: Assertion `idx < size()' failed.
Aborted (core dumped)
```

I'm stuck, i should look deeper into this or leave it for now .. maybe i'll try compiling again whenever you create a (major) update. Do you know if c++23 is partially implemented, as i wrote ? It's a pitty ..

**Comment by JoostHouben** [comment]
@tissatussa if you share your code in a fork on your profile I can take a quick look.

> Do you know if c++23 is partially implemented

Yes. I believe all the necessary C++23 features are implemented in the latest release of GCC (14.2). They're also implemented in the latest pre-release preview of clang, but not in the most recent public release (18.1). See here for a quick test case: https://godbolt.org/z/eb4zPn8xE

**Comment by tissatussa** [comment]
Today i also encountered the "Compiler Explorer" website https://godbolt.org : it's really nice ! Some of my code (parts) run fine here, using clang++-17, but NOT on my OS, and that's weird ..

I have GCC-13, but indeed it couldn't compile your source .. i didn't know 14.2 exists, i will try !

> if you share your code in a fork on your profile I can take a quick look

i don't know how to do that, sorry .. so i will post a ZIP here : [Euwe-v1.0.1-fixed-for-my-linux-but-still-buggy.zip](https://github.com/user-attachments/files/16640746/Euwe-v1.0.1-fixed-for-my-linux-but-still-buggy.zip) In the Makefile i skipped the (Google) test parts, they gave new problems and i think this functionality is not really needed, am i right ? If you use a tool like Meld (GUI for 'diff') my changes are easy to see.

**Comment by tissatussa** [comment]
@JoostHouben i really like to solve this, creating code for Linux .. and learn. Thanks for your feedback, i'll stay tuned.

**Comment by JoostHouben** [comment]
Yeah, you're not forwarding the correct output stream to your `myPrintLN` function.

This works: https://godbolt.org/z/oE5ssn36o

`myPrintLN` needs to take a stream argument to specify where the output should go:
```
template<typename... Args>
void myPrintLN(std::ostream& out, const std::format_string<Args...> fmt, Args&&... args) {
    out << std::format(fmt, std::forward<Args>(args)...) << '\n';
}
```

And then you need to pass the correct stream to the print function in `writeUci` and its siblings:

```
template <typename... Args>
void writeUci(const std::format_string<Args...> fmt, Args&&... args) {
    myPrintLN(out_, fmt, std::forward<Args>(args)...);
}

template <typename... Args>
void writeDebug(const std::format_string<Args...> fmt, Args&&... args) {
    if (debugMode_) {
        myPrintLN(out_, "info string {}", std::format(fmt, std::forward<Args>(args)...));
    } else {
        myPrintLN(debug_, "[DEBUG] {}", std::format(fmt, std::forward<Args>(args)...));
    }
}

template <typename... Args>
void writeDebugNonUci(
        const std::format_string<Args...> fmt, Args&&... args) {
    myPrintLN(debug_, "[DEBUG] {}", std::format(fmt, std::forward<Args>(args)...));
}
```

**Comment by tissatussa** [comment]
Great ! I will look into this and change my code accordingly.

Later, i'll be out now :-)

**Comment by tissatussa** [comment]
i managed to install gcc-14.2.0
you didn't comment on me skipping Google Test, is it needed ?  What about my Makefile, does it suit gcc, or should some parameters be adjusted for that ? I like good Makefiles ..

**Comment by tissatussa** [comment]
i successfully compiled v1.0.1 (commit 6f51934) on Linux with gcc 14.2.0 without any hassle .. i only had to adjust 2 minor things in the code :

in `UciFrontEnd.cpp` the function `std::round()` is used (once) and i had to change this to `round()`, adding `#include <cmath>` on top of the file.
in `Main.cpp` i removed the path `chess-engine-lib/` from the 4 include file declarations, because my Makefile includes these folders :

```
# Compiler and flags
CXX = /usr/local/gcc-14.2.0/bin/g++-14.2.0
CXXFLAGS = -std=c++23 -Wall -Wextra -pedantic -I./chess-engine-lib -mbmi2 -mavx2
LDFLAGS = -pthread -lm

# Directories
SRCDIR = chess-engine
LIBDIR = chess-engine-lib
# TESTDIR = tests
BUILDDIR = build

# Source files
LIB_SRCS = $(wildcard $(LIBDIR)/*.cpp)
LIB_OBJS = $(patsubst $(LIBDIR)/%.cpp,$(BUILDDIR)/%.o,$(LIB_SRCS))

ENGINE_SRCS = $(SRCDIR)/Main.cpp
ENGINE_OBJS = $(patsubst $(SRCDIR)/%.cpp,$(BUILDDIR)/%.o,$(ENGINE_SRCS))

# TEST_SRCS = $(wildcard $(TESTDIR)/*.cpp)
# TEST_OBJS = $(patsubst $(TESTDIR)/%.cpp,$(BUILDDIR)/test_%.o,$(TEST_SRCS))

# Output binary
ENGINE_BIN = euwe-engine
# TEST_BIN = euwe-tests

# Default target
# all: $(ENGINE_BIN) $(TEST_BIN)
all: $(ENGINE_BIN)

# Build the engine binary
$(ENGINE_BIN): $(LIB_OBJS) $(ENGINE_OBJS)
    $(CXX) $(LIB_OBJS) $(ENGINE_OBJS) -o $@ $(LDFLAGS)

# # Build the test binary
# $(TEST_BIN): $(LIB_OBJS) $(TEST_OBJS)
#   $(CXX) $(LIB_OBJS) $(TEST_OBJS) -o $@ $(LDFLAGS)

# Compile object files
$(BUILDDIR)/%.o: $(LIBDIR)/%.cpp | $(BUILDDIR)
    $(CXX) $(CXXFLAGS) -c $< -o $@

$(BUILDDIR)/%.o: $(SRCDIR)/%.cpp | $(BUILDDIR)
    $(CXX) $(CXXFLAGS) -c $< -o $@

# $(BUILDDIR)/test_%.o: $(TESTDIR)/%.cpp | $(BUILDDIR)
#   $(CXX) $(CXXFLAGS) -c $< -o $@

# Create build directory
$(BUILDDIR):
    mkdir -p $(BUILDDIR)

# Clean up
clean:
#   rm -rf $(BUILDDIR) $(ENGINE_BIN) $(TEST_BIN)
    rm -rf $(BUILDDIR) $(ENGINE_BIN)
```

The script uses a folder `build/`.
The Google Test parts are skipped for now.

It creates a 1.2 Mb binary on my notebook.
The engine can play in CuteChess GUI, i did 4 5m+3s games, but in ALL games Euwe disconnected .. sometimes at move 15, sometimes at move 4 .. here's the CuteChess Log tail :

```
euwe-engine-v1.0.1-compiled-HP-commit-6f51934: chess-engine-lib/GameState.cpp:689: Piece GameState::makeSinglePieceMove(const Move&): Assertion `move.to == enPassantTarget_' failed.
```

btw. it also logs all your [DEBUG] messages. I didn't find a way in CuteChess GUI (v1.3.1) to set the `stderrFile` path, but it can be done by hand in `~/.config/cutechess.json` for each configured engine.

**Comment by JoostHouben** [comment]
GTest is only needed for running the tests, not for running the engine itself.

I took a quick glance at the Makefile. It looks fine for the most part, but you probably want to make sure that you're compiling with optimizations enabled (-O2 or similar).

Hmm, that crash you mentioned is surprising. I'll look into it. Would you mind opening a new issue for that and including the full sequence of moves leading up to that position (ideally, the `position ...` string from the UCI input)? That will help me debug.

**Comment by tissatussa** [comment]
> (Makefile) looks fine for the most part, but you probably want to make sure that you're compiling with optimizations enabled (-O2 or similar)

i think the Makefile is simple and OK, just my basic understanding.
indeed optimisation parameters can be added, but first i wanted to make it work - optimisations can better be added after solving the disconnecting.

yes, i will open a new Issue for this disconnection bug and supply logs.
